### DISCUSSION TITLE
SML issue with OBIS ASCII on negative current values

### DISCUSSION THREAD

**Discussion by DaGo1120** [arendst/Tasmota#20593] [Q&A] [answered]
Hi! I Have an issue with some values, transmitted via the SML OBIS ASCII interface.
Dumped values look like this:
`15:40:33.963 : /LOG5LK13BE803039`
`15:40:33.964 : 1-0:96.1.0*255(001LOG0065200716)`
`15:40:34.013 : 1-0:1.8.0*255(015794.0248*kWh)`
`15:40:34.063 : 1-0:2.8.0*255(000000.6452*kWh)`
`15:40:34.065 : 1-0:16.7.0*255(000777*W)`
`15:40:34.113 : 1-0:32.7.0*255(228.7*V)`
`15:40:34.162 : 1-0:52.7.0*255(227.2*V)`
`15:40:34.164 : 1-0:72.7.0*255(229.6*V)`
`15:40:34.212 : 1-0:31.7.0*255(002.89*A)`
`15:40:34.214 : 1-0:51.7.0*255(000.63*A)`
`15:40:34.263 : 1-0:71.7.0*255(000.33*A)`
`15:40:34.265 : 1-0:81.7.1*255(125*deg)`
`15:40:34.313 : 1-0:81.7.2*255(242*deg)`
`15:40:34.314 : 1-0:81.7.4*255(018*deg)`
`15:40:34.363 : 1-0:81.7.15*255(045*deg)`
`15:40:34.412 : 1-0:81.7.26*255(052*deg)`
`15:40:34.414 : 1-0:14.7.0*255(49.9*Hz)`
`15:40:34.462 : 1-0:1.8.0*96(00025.3*kWh)`
`15:40:34.464 : 1-0:1.8.0*97(00192.0*kWh)`
`15:40:34.513 : 1-0:1.8.0*98(00958.1*kWh)`
`15:40:34.514 : 1-0:1.8.0*99(03946.5*kWh)`
`15:40:34.563 : 1-0:1.8.0*100(15794.0*kWh)`
`15:40:34.613 : 1-0:0.2.0*255(ver.03,432F,20170504)`
`15:40:34.615 : 1-0:96.90.2*255(0F66)`
`15:40:34.663 : 1-0:97.97.0*255(00000000)`
`15:40:34.664 : !`

And this is my script:
`>D`
`>B`
`=>sensor53 r`
`>M 1`
`+1,4,o,0,9600,LK13BE`

`1,1-0:1.8.0*255(@1,Gesamtverbrauch,KWh,Etotal,4`
`1,1-0:1.8.0*96(@1,Verbrauch 1 Tag,KWh,Etotal_1d,1`
`1,1-0:1.8.0*97(@1,Verbrauch 7 Tage,KWh,Etotal_7d,1`
`1,1-0:1.8.0*98(@1,Verbrauch 30 Tage,KWh,Etotal_30d,1`
`1,1-0:1.8.0*99(@1,Verbrauch 365 Tage,KWh,Etotal_365d,1`
`1,1-0:2.8.0*255(@1,Gesamteinspeisung,KWh,Etotal_back,4`
`1,1-0:16.7.0*255(@1,Netz Leistung,W,power,17`
`1,1-0:14.7.0*255(@1,Netz Frequenz,Hz,frequency,1`
`1,1-0:31.7.0*255(@1,Strom L1,A,current_L1,2`
`1,1-0:32.7.0*255(@1,Spannung L1,V,voltage_L1,1`
`1.1-0:81.7.4*255(@1,Phasenwinkel L1,deg,phase_L1,0`
`1,1-0:51.7.0*255(@1,Strom L2,A,current_L2,2`
`1,1-0:52.7.0*255(@1,Spannung L2,V,voltage_L2,1`
`1.1-0:81.7.15*255(@1,Phasenwinkel L2,deg,phase_L2,0`
`1,1-0:71.7.0*255(@1,Strom L3,A,current_L3,2`
`1,1-0:72.7.0*255(@1,Spannung L3,V,voltage_L3,1`
`1.1-0:81.7.26*255(@1,Phasenwinkel L3,deg,phase_L3,0`
`#`

I'm calculating the power of the L1 phase with this formula:
power_L1 = current_L1 * voltage_L1 * cos(phase_L1 * PI / 180)

In most cases, I get valid values. But if the power turns negative (I source power to the grid), my calculation still delivers positive values.
Is there something wrong with my script or my formula?
Unfortunately I can't reproduce the log above during this situation, due to the lack of solar energy. But I hope, someone can help me anyway!

I read about the M,=so1 option but did not understand it. If I'm right, this only affects binary data and not ASCII data...

**Comment by DaGo1120** [comment]
My Tasmota version is 12.0.4

**Comment by gemu2015** [comment]
not enough info.
does the SML webgui show negative numbers.
where do you do your calculations ?

**Comment by DaGo1120** [comment]
Sorry, you are right!
The webgui values for I1, I2 and I3 are also absolute (never negative). I transmit everything via MQTT and do the calculations externally (IoBroker)

**Reply by gemu2015** [reply]
Ok then we need a dump when you feed in to determine weather the meter does report negative values

**Reply by DaGo1120** [reply]
Ok, here is a dump:
```
10:11:03.046 : /LOG5LK13BE803039
10:11:03.048 : 1-0:96.1.0*255(001LOG0065200716)
10:11:03.097 : 1-0:1.8.0*255(015855.5496*kWh)
10:11:03.147 : 1-0:2.8.0*255(000001.4907*kWh)
10:11:03.149 : 1-0:16.7.0*255(-000043*W)
10:11:03.197 : 1-0:32.7.0*255(229.6*V)
10:11:03.199 : 1-0:52.7.0*255(228.1*V)
10:11:03.247 : 1-0:72.7.0*255(230.8*V)
10:11:03.249 : 1-0:31.7.0*255(001.16*A)
10:11:03.297 : 1-0:51.7.0*255(000.63*A)
10:11:03.346 : 1-0:71.7.0*255(000.25*A)
10:11:03.349 : 1-0:81.7.1*255(124*deg)
10:11:03.397 : 1-0:81.7.2*255(241*deg)
10:11:03.398 : 1-0:81.7.4*255(048*deg)
10:11:03.447 : 1-0:81.7.15*255(044*deg)
10:11:03.448 : 1-0:81.7.26*255(058*deg)
10:11:03.497 : 1-0:14.7.0*255(49.9*Hz)
10:11:03.546 : 1-0:1.8.0*96(00024.9*kWh)
10:11:03.548 : 1-0:1.8.0*97(00166.6*kWh)
10:11:03.596 : 1-0:1.8.0*98(00949.6*kWh)
10:11:03.598 : 1-0:1.8.0*99(03984.4*kWh)
10:11:03.647 : 1-0:1.8.0*100(15855.5*kWh)
10:11:03.696 : 1-0:0.2.0*255(ver.03,432F,20170504)
10:11:03.698 : 1-0:96.90.2*255(0F66)
10:11:03.746 : 1-0:97.97.0*255(00000000)
10:11:03.748 : !
```
or an other:
```
10:16:04.048 : /LOG5LK13BE803039
10:16:04.050 : 1-0:96.1.0*255(001LOG0065200716)
10:16:04.099 : 1-0:1.8.0*255(015855.5496*kWh)
10:16:04.148 : 1-0:2.8.0*255(000001.4962*kWh)
10:16:04.151 : 1-0:16.7.0*255(-000095*W)
10:16:04.199 : 1-0:32.7.0*255(230.6*V)
10:16:04.201 : 1-0:52.7.0*255(227.6*V)
10:16:04.249 : 1-0:72.7.0*255(230.8*V)
10:16:04.250 : 1-0:31.7.0*255(001.19*A)
10:16:04.299 : 1-0:51.7.0*255(000.63*A)
10:16:04.348 : 1-0:71.7.0*255(000.20*A)
10:16:04.351 : 1-0:81.7.1*255(117*deg)
10:16:04.398 : 1-0:81.7.2*255(242*deg)
10:16:04.401 : 1-0:81.7.4*255(037*deg)
10:16:04.449 : 1-0:81.7.15*255(046*deg)
10:16:04.450 : 1-0:81.7.26*255(061*deg)
10:16:04.499 : 1-0:14.7.0*255(49.9*Hz)
10:16:04.500 : 1-0:1.8.0*96(00024.9*kWh)
10:16:04.549 : 1-0:1.8.0*97(00166.6*kWh)
10:16:04.598 : 1-0:1.8.0*98(00949.6*kWh)
10:16:04.600 : 1-0:1.8.0*99(03984.4*kWh)
10:16:04.649 : 1-0:1.8.0*100(15855.5*kWh)
10:16:04.698 : 1-0:0.2.0*255(ver.03,432F,20170504)
10:16:04.701 : 1-0:96.90.2*255(0F66)
10:16:04.748 : 1-0:97.97.0*255(00000000)
10:16:04.750 : !
```
The total power is negative, so at least one of the currents - in this case the line
`10:11:03.249 : 1-0:31.7.0*255(001.16*A)`
should be negative.

**Reply by gemu2015** [reply, marked as answer]
Ok then this is an issue with your meter. You will not know which phase is negative. There is now way to evaluate this from the meter data.

**Reply by DaGo1120** [reply]
Ok, thank you. I try to solve this outside of Tasmota in my calculation.

**Reply by DaGo1120** [reply]
Ok is there a way, to to the following via the Tasmota scripting engine?
I mean something like
```
power_L1 = current_L1 * voltage_L1 * cos(phase_L1 * PI / 180)
power_L2 = current_L2 * voltage_L2 * cos(phase_L2 * PI / 180)
power_L3 = current_L3 * voltage_L3 * cos(phase_L3 * PI / 180)

if (power_total_from_meter < (power_L1 + power_L2 + power_L3) then
    power_L1 = power_L1 * -1
fi
```
I1 is the only phase, I source power. So I can be sure, L1 is the only phase, which can be negative.

Edit: I have found this and will do my best to implement it. Wish me luck! :-)
https://tasmota.github.io/docs/Scripting-Language/

**Reply by gemu2015** [reply]
you may easily do your above calculations in the sml script itself.
keep in mind that you must compile with
#define USE_ANGLE_FUNC
to use cos(x)
